The new version causes build issues

Hello there,

I'm building a musl based rust build container with some pre-compiled C libraries, including libxml2.
All previous versions seem to work perfectly and build without issues.

But since the release of the new version this now breaks.
Not specifically sure why this is happening though, as i have not yet look into it deeply.
But maybe somebody here knows a way to fix it, or see what the issues more quickly :) 

The error can be viewed here:
https://github.com/BlackDex/rust-musl/actions/runs/14512752243/job/40714970804#step:33:324

Any pointers would be great.

Ok, it looked like i used a `make` command which did not installed the `.c` and `.h` files.

I now get the following error:
```
error[E0308]: mismatched types
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/c_helpers.rs:112:57
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |                                                    |
    |                                                    arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                  found fn item `unsafe extern "C" fn(_, *mut _xmlError) {_ignoreInvalidTagsErrorFunc}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {_ignoreInvalidTagsErrorFunc}` due to the type of the argument passed
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/c_helpers.rs:112:52
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ^^^^^---------------------------^
    |                                                         |
    |                                                         this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

error[E0308]: mismatched types
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/tree/node.rs:418:14
    |
418 |     unsafe { xmlNodeSetContent(self.node_ptr_mut()?, c_content.as_bytes().as_ptr()) }
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- help: consider using a semicolon here: `;`
    |              |
    |              expected `()`, found `i32`

error[E0308]: mismatched types
  --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/schemas/parser.rs:82:14
   |
82 |         Some(common::structured_error_handler),
   |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
   |         |
   |         arguments to this enum variant are incorrect
   |
   = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                 found fn item `unsafe extern "C" fn(_, *mut _xmlError) {structured_error_handler}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {structured_error_handler}` due to the type of the argument passed
  --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/schemas/parser.rs:82:9
   |
82 |         Some(common::structured_error_handler),
   |         ^^^^^--------------------------------^
   |              |
   |              this argument influences the type of `Some`
note: tuple variant defined here
  --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

error[E0308]: mismatched types
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/schemas/validation.rs:104:14
    |
104 |         Some(common::structured_error_handler),
    |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |         |
    |         arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                  found fn item `unsafe extern "C" fn(_, *mut _xmlError) {structured_error_handler}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {structured_error_handler}` due to the type of the argument passed
   --> /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/schemas/validation.rs:104:9
    |
104 |         Some(common::structured_error_handler),
    |         ^^^^^--------------------------------^
    |              |
    |              this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

For more information about this error, try `rustc --explain E0308`.
error: could not compile `libxml` (lib) due to 4 previous errors

Caused by:
  process didn't exit successfully: `CARGO=/usr/local/rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo CARGO_CRATE_NAME=libxml CARGO_MANIFEST_DIR=/usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4 CARGO_MANIFEST_PATH=/usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/Cargo.toml CARGO_PKG_AUTHORS='Andreas Franzén <andreas@devil.se>:Deyan Ginev <deyan.ginev@gmail.com>:Jan Frederik Schaefer <j.schaefer@jacobs-university.de>' CARGO_PKG_DESCRIPTION='A Rust wrapper for libxml2 - the XML C parser and toolkit developed for the Gnome project' CARGO_PKG_HOMEPAGE='' CARGO_PKG_LICENSE=MIT CARGO_PKG_LICENSE_FILE='' CARGO_PKG_NAME=libxml CARGO_PKG_README=README.md CARGO_PKG_REPOSITORY='https://github.com/KWARC/rust-libxml' CARGO_PKG_RUST_VERSION='' CARGO_PKG_VERSION=0.3.4 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_VERSION_MINOR=3 CARGO_PKG_VERSION_PATCH=4 CARGO_PKG_VERSION_PRE='' LD_LIBRARY_PATH='/home/rust/src/target/debug/deps:/usr/local/rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib' OUT_DIR=/home/rust/src/target/x86_64-unknown-linux-musl/debug/build/libxml-0b891eab3728ee47/out /usr/local/rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --crate-name libxml --edition=2021 /usr/local/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/libxml-0.3.4/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=268 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values())' -C metadata=808ef9371f6be4e5 -C extra-filename=-9aa406fe2004f28c --out-dir /home/rust/src/target/x86_64-unknown-linux-musl/debug/deps --target x86_64-unknown-linux-musl -C linker=x86_64-unknown-linux-musl-ld -L dependency=/home/rust/src/target/x86_64-unknown-linux-musl/debug/deps -L dependency=/home/rust/src/target/debug/deps --extern libc=/home/rust/src/target/x86_64-unknown-linux-musl/debug/deps/liblibc-eb5d845088983b4e.rmeta --cap-lints warn -Clink-arg=-s -L native=/usr/local/musl/lib -l static=xml2 -l m` (exit status: 1)
warning: build failed, waiting for other jobs to finish...
```

Also, not looked deeply into this yet.

I just run `cargo +nightly clippy` on the codebase, and it returns an error regarding those items too.

```
error[E0308]: mismatched types
   --> src/c_helpers.rs:112:57
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected fn pointer, found fn item
    |                                                    |
    |                                                    arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(*mut libc::c_void, _)`
                  found fn item `unsafe extern "C" fn(libc::c_void, _) {c_helpers::_ignoreInvalidTagsErrorFunc}`
help: the type constructed contains `unsafe extern "C" fn(libc::c_void, *mut bindings::_xmlError) {c_helpers::_ignoreInvalidTagsErrorFunc}` due to the type of the argument passed
   --> src/c_helpers.rs:112:52
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ^^^^^---------------------------^
    |                                                         |
    |                                                         this argument influences the type of `Some`
note: tuple variant defined here
   --> /home/mathijs/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:580:5
    |
580 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `libxml` (lib) due to 1 previous error
```

I'm using:
```
cargo +nightly -vV
cargo 1.88.0-nightly (864f74d4e 2025-04-11)
release: 1.88.0-nightly
commit-hash: 864f74d4eadcaea3eeda37a2e7f4d34de233d51e
commit-date: 2025-04-11
host: x86_64-unknown-linux-gnu
libgit2: 1.9.0 (sys:0.20.0 vendored)
libcurl: 8.12.1-DEV (sys:0.4.80+curl-8.12.1 vendored ssl:OpenSSL/3.4.1)
ssl: OpenSSL 3.4.1 11 Feb 2025
os: Ubuntu 22.4.0 (jammy) [64-bit]
```

I did a new clone, and for some reason it doesn't complain about the previous items right now, and only the same you also see.

The only pending items which prevent me from using 0.3.4 are these:

```
error[E0308]: mismatched types
   --> .artifacts/rust-libxml/src/c_helpers.rs:112:57
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |                                                    |
    |                                                    arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                  found fn item `unsafe extern "C" fn(_, *mut _xmlError) {_ignoreInvalidTagsErrorFunc}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {_ignoreInvalidTagsErrorFunc}` due to the type of the argument passed
   --> .artifacts/rust-libxml/src/c_helpers.rs:112:52
    |
112 |     xmlSetStructuredErrorFunc(ctxt as *mut c_void, Some(_ignoreInvalidTagsErrorFunc));
    |                                                    ^^^^^---------------------------^
    |                                                         |
    |                                                         this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

error[E0308]: mismatched types
  --> .artifacts/rust-libxml/src/schemas/parser.rs:82:14
   |
82 |         Some(common::structured_error_handler),
   |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
   |         |
   |         arguments to this enum variant are incorrect
   |
   = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                 found fn item `unsafe extern "C" fn(_, *mut _xmlError) {structured_error_handler}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {structured_error_handler}` due to the type of the argument passed
  --> .artifacts/rust-libxml/src/schemas/parser.rs:82:9
   |
82 |         Some(common::structured_error_handler),
   |         ^^^^^--------------------------------^
   |              |
   |              this argument influences the type of `Some`
note: tuple variant defined here
  --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

error[E0308]: mismatched types
   --> .artifacts/rust-libxml/src/schemas/validation.rs:104:14
    |
104 |         Some(common::structured_error_handler),
    |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |         |
    |         arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, *const _xmlError)`
                  found fn item `unsafe extern "C" fn(_, *mut _xmlError) {structured_error_handler}`
help: the type constructed contains `unsafe extern "C" fn(*mut c_void, *mut _xmlError) {structured_error_handler}` due to the type of the argument passed
   --> .artifacts/rust-libxml/src/schemas/validation.rs:104:9
    |
104 |         Some(common::structured_error_handler),
    |         ^^^^^--------------------------------^
    |              |
    |              this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/05f9846f893b09a1be1fc8560e33fc3c815cfecb/library/core/src/option.rs:580:5

For more information about this error, try `rustc --explain E0308`.
error: could not compile `libxml` (lib) due to 3 previous errors
```

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -27,3 +27,35 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
+
+  test-libxml2_12:
+    name: With libxml 2.12
+    runs-on: ubuntu-latest
+    steps:
+      - name: install dependencies
+        uses: ryankurte/action-apt@v0.2.0
+        with:
+          packages: "libpython3-dev"
+      - uses: actions/checkout@v2
+      - name: Install libxml 2.12 by hand
+        run: |
+          wget https://download.gnome.org/sources/libxml2/2.12/libxml2-2.12.9.tar.xz
+          tar xf libxml2-2.12.9.tar.xz
+          cd libxml2-2.12.9
+          ./configure
+          make
+          cp libxml-2.0-uninstalled.pc libxml-2.0.pc
+          cd include
+          ln -s . libxml2
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+      - name: run tests
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+        env:
+          PKG_CONFIG_PATH: "./libxml2-2.12.9"
+

--- a/build.rs
+++ b/build.rs
@@ -1,8 +1,13 @@
 use std::{env, fs, path::{Path, PathBuf}};
 
+struct ProbedLib {
+  version: String,
+  include_paths: Vec<PathBuf>,
+}
+
 /// Finds libxml2 and optionally return a list of header
 /// files from which the bindings can be generated.
-fn find_libxml2() -> Option<Vec<PathBuf>> {
+fn find_libxml2() -> Option<ProbedLib> {
   #![allow(unreachable_code)] // for platform-dependent dead code
 
   if let Ok(ref s) = std::env::var("LIBXML2") {
@@ -40,7 +45,10 @@ fn find_libxml2() -> Option<Vec<PathBuf>> {
       let lib = pkg_config::Config::new()
         .probe("libxml-2.0")
         .expect("Couldn't find libxml2 via pkg-config");
-      return Some(lib.include_paths)
+      return Some(ProbedLib {
+        include_paths: lib.include_paths,
+        version: lib.version,
+      })
     }
 
     #[cfg(windows)]
@@ -75,13 +83,18 @@ fn generate_bindings(header_dirs: Vec<PathBuf>, output_path: &Path) {
 
 fn main() {
   let bindings_path = PathBuf::from(env::var_os("OUT_DIR").unwrap()).join("bindings.rs");
-  if let Some(header_dirs) = find_libxml2() {
+  // declare availability of config variable (without setting it)
+  println!("cargo::rustc-check-cfg=cfg(libxml_older_than_2_12)");
+
+  if let Some(probed_lib) = find_libxml2() {
     // if we could find header files, generate fresh bindings from them
-    generate_bindings(header_dirs, &bindings_path);
+    generate_bindings(probed_lib.include_paths, &bindings_path);
   } else {
     // otherwise, use the default bindings on platforms where pkg-config isn't available
     fs::copy(PathBuf::from("src/default_bindings.rs"), bindings_path)
       .expect("Failed to copy the default bindings to the build directory");
+    // for now, assume that the library is older than 2.12, because that's what those bindings are computed with
+    println!("cargo::rustc-cfg=libxml_older_than_2_12");
   }
 }
 

--- a/build.rs
+++ b/build.rs
@@ -89,6 +89,15 @@ fn main() {
   if let Some(probed_lib) = find_libxml2() {
     // if we could find header files, generate fresh bindings from them
     generate_bindings(probed_lib.include_paths, &bindings_path);
+    // and expose the libxml2 version to the code
+    let version_parts: Vec<i32> = probed_lib.version.split('.')
+      .map(|part| part.parse::<i32>().unwrap_or(-1)).collect();
+    let older_than_2_12 = version_parts.len() > 1 && (version_parts[0] < 2 ||
+        version_parts[0] == 2 && version_parts[1] < 12);
+    println!("cargo::rustc-check-cfg=cfg(libxml_older_than_2_12)");
+    if older_than_2_12 {
+      println!("cargo::rustc-cfg=libxml_older_than_2_12");
+    }
   } else {
     // otherwise, use the default bindings on platforms where pkg-config isn't available
     fs::copy(PathBuf::from("src/default_bindings.rs"), bindings_path)

--- a/src/c_helpers.rs
+++ b/src/c_helpers.rs
@@ -97,7 +97,17 @@ pub fn xmlNodeGetName(cur: xmlNodePtr) -> *const c_char {
 }
 
 // dummy function: no debug output at all
+#[cfg(libxml_older_than_2_12)]
 unsafe extern "C" fn _ignoreInvalidTagsErrorFunc(_user_data: *mut c_void, error: xmlErrorPtr) {
+  unsafe {
+    if !error.is_null() && (*error).code as u32 == xmlParserErrors_XML_HTML_UNKNOWN_TAG {
+      // do not record invalid, in fact (out of despair) claim we ARE well-formed, when a tag is invalid.
+      HACKY_WELL_FORMED = true;
+    }
+  }
+}
+#[cfg(not(libxml_older_than_2_12))]
+unsafe extern "C" fn _ignoreInvalidTagsErrorFunc(_user_data: *mut c_void, error: *const xmlError) {
   unsafe {
     if !error.is_null() && (*error).code as u32 == xmlParserErrors_XML_HTML_UNKNOWN_TAG {
       // do not record invalid, in fact (out of despair) claim we ARE well-formed, when a tag is invalid.

--- a/src/error.rs
+++ b/src/error.rs
@@ -66,7 +66,7 @@ impl StructuredError {
   /// This function copies data from the memory `error_ptr` but does not deallocate
   /// the error. Depending on the context in which this function is used, you may
   /// need to take additional steps to avoid a memory leak.
-  pub unsafe fn from_raw(error_ptr: *mut bindings::xmlError) -> Self {
+  pub unsafe fn from_raw(error_ptr: *const bindings::xmlError) -> Self {
     let error = *error_ptr;
     let message = StructuredError::ptr_to_string(error.message);
     let level = XmlErrorLevel::from_raw(error.level);

--- a/src/schemas/common.rs
+++ b/src/schemas/common.rs
@@ -9,7 +9,18 @@ use std::ffi::c_void;
 
 /// Provides a callback to the C side of things to accumulate xmlErrors to be
 /// handled back on the Rust side.
+#[cfg(libxml_older_than_2_12)]
 pub unsafe extern "C" fn structured_error_handler(ctx: *mut c_void, error: bindings::xmlErrorPtr) {
+  assert!(!ctx.is_null());
+  let errlog = unsafe { &mut *{ ctx as *mut Vec<StructuredError> } };
+
+  let error = unsafe { StructuredError::from_raw(error) };
+
+  errlog.push(error);
+}
+
+#[cfg(not(libxml_older_than_2_12))]
+pub unsafe extern "C" fn structured_error_handler(ctx: *mut c_void, error: *const bindings::xmlError) {
   assert!(!ctx.is_null());
   let errlog = unsafe { &mut *{ ctx as *mut Vec<StructuredError> } };
 

